src/main/scala/Validator.scala

```diff
@@ -1,18 +1,103 @@
 package CSVValidation
 
+import org.apache.commons.csv.{CSVFormat, CSVParser, CSVRecord}
+
+import java.io.File
+import java.net.URI
+import java.nio.charset.StandardCharsets
+import scala.jdk.CollectionConverters.IterableHasAsScala
+
 // This is only a skeleton for the core Validator class to be created. It is planned to accept the source csv and
 // schema and return errors and warnings collection when validate method is invoked. Changes to this signature can be
 // made in the due course or as per requirements
-class Validator(val filePath: String) {
+class Validator(var tableCsvFile: URI, sourceUri: String = "") {
+  val isCsvHeader = true
+  var warnings: Array[ErrorMessage] = Array()
+  var errors: Array[ErrorMessage] = Array()
+  var schema: Option[TableGroup] = None
+  var source: String = "" // Define how this is set, will it always be string?
+  val csvHeaderExpected = true
+
   def validate(): Either[String, Array[String]] = {
-    val result = Schema.loadMetadataAndValidate(filePath)
-    result match {
-      case Right(tableGroup) =>
+    val mayBeTableGroupObject = Schema.loadMetadataAndValidate(tableCsvFile)
+    mayBeTableGroupObject match {
+      case Right(tableGroup) => {
+        schema = Some(tableGroup)
+        if (sourceUri.isEmpty) {
+          fetchSchemaTables(tableGroup)
+        }
         Right(tableGroup.warnings.map(w => processWarnings(w)))
+      }
       case Left(errorMessage) => Left(errorMessage)
     }
   }
 
+  def fetchSchemaTables(schema: TableGroup) = {
+    for (tableUrl <- schema.tables.keys) {
+      val tableUri = new URI(tableUrl)
+      if (tableUri.getScheme == "file") { // tableUrl is a filename in this case, so check if it exists
+        val tableCsvFile = new File(tableUri)
+        if (!tableCsvFile.exists) {
+          throw new MetadataError(
+            "CSV NOT FOUND"
+          ) // Change this to an error returned to the CLI
+        }
+        readCSV(tableUri)
+      } else readCSV(tableUri)
+      // Call validate csv function
+    }
+  }
+
+  def readCSV(tableUri: URI) = {
+    val parser = if (tableUri.getScheme == "file") {
+      CSVParser
+        .parse(
+          new File(tableUri),
+          StandardCharsets.UTF_8,
+          CSVFormat.DEFAULT
+        )
+    } else {
+      CSVParser.parse(
+        tableUri.toURL,
+        StandardCharsets.UTF_8,
+        CSVFormat.DEFAULT
+      )
+    }
+    for (row <- parser.asScala) {
+      if (row.getRecordNumber == 1 && csvHeaderExpected) {
+        validateHeader(row, tableUri)
+      } else {
+        if (row.size == 0) {
+          warnings :+= ErrorMessage(
+            "Blank rows",
+            "structure",
+            row.getRecordNumber.toString,
+            "",
+            "",
+            ""
+          )
+        }
+        schema match {
```

Comment:
This match statement should be outside of the for loop - I suspect the cost of it will add up quite quickly.